<a href="https://colab.research.google.com/github/arpitp2912/Deep_Learning_Assignments/blob/master/Numpy_Implementation/Numpy_Implementation_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Numpy Implementation of 1 Hidden Layer Neural Network

Let’s say we have a problem where we want to predict output given a set of inputs and outputs as training example like:

| Input 1 | Input 2 | Input 3 | Output
| --- | --- | --- | --- |
| 0 | 1 | 1 | 1
| 1 | 0 | 0 | 0
| 1 | 0 | 1 | 1


Now, we want to predict the output  of the inputs : **[1,0,1]**

The training process consists of the following steps:


*   **Forward Propagation:**

> Take the inputs, multiply by the weights (just use random numbers as weights)
Let Y = WiIi = W1I1+W2I2+W3I3
Pass the result through a sigmoid formula to calculate the neuron’s output. The Sigmoid function is used to normalise the result between 0 and 1:
1/1 + e-y


*  **Back Propagation**

> Calculate the error i.e the difference between the actual output and the expected output. Depending on the error, adjust the weights by multiplying the error with the input and again with the gradient of the Sigmoid curve:
Weight += Error Input Output (1-Output) ,here Output (1-Output) is derivative of sigmoid curve.


In [0]:
# Import Library
import numpy as np

In [0]:
class NeuralNet(object):
    def __init__(self):
      # Assign random weights to a 3 x 1 matrix
      self.synaptic_weights = 2 * random.random((3, 1)) - 1

    # The Sigmoid function
    def __sigmoid(self, x):
        return 1 / (1 + exp(-x))

    # The Gradient of Sigmoid function.
    def __sigmoid_derivative(self, x):
        return x * (1 - x)

    # Train the neural network and adjust the weights each time.
    def train(self, inputs, outputs, training_iterations):
        for iteration in range(training_iterations):
        
            # Pass the training set through the network.
            output = self.learn(inputs)

            # Calculate the error
            error = outputs - output

            # Adjust the weights
            adjustment = dot(inputs.T, error * self.__sigmoid_derivative(output))
            self.synaptic_weights += adjustment

    # Learning of Neural Network
    def learn(self, inputs):
        return self.__sigmoid(dot(inputs, self.synaptic_weights))


if __name__ == "__main__":

    #Initialize the network
    neural_network = NeuralNet()
    
    # The training set.
    inputs = array([[0, 1, 1], [1, 0, 0], [1, 0, 1]])
    outputs = array([[1, 0, 1]]).T
    
    # Train the network (10 epochs)
    neural_network.train(inputs, outputs, 10)

    # Test the neural network with a test example.
    print('Result after 10 epochs: ',neural_network.learn(array([1, 0, 1])))
    
    # Train the network (100 epochs)
    neural_network.train(inputs, outputs, 100)

    # Test the neural network with a test example.
    print('Result after 100 epochs: ',neural_network.learn(array([1, 0, 1])))
    
    # Train the network (1000 epochs)
    neural_network.train(inputs, outputs, 1000)

    # Test the neural network with a test example.
    print('Result after 1000 epochs: ',neural_network.learn(array([1, 0, 1])))
    
    # Train the network (10000 epochs)
    neural_network.train(inputs, outputs, 10000)

    # Test the neural network with a test example.
    print('Result after 10000 epochs: ',neural_network.learn(array([1, 0, 1])))



Result after 10 epochs:  [0.63290706]
Result after 100 epochs:  [0.88349355]
Result after 1000 epochs:  [0.96781691]
Result after 10000 epochs:  [0.99030646]


**Results:**

> After 10 iteration, our network predicts the value to be 0.632. It does not look good as the asnwer should be 1.



> If we increase the number of iterations to 100, we get 0.88349355. Our network is getting smarter! Subsequently, for 10000 iterations we get 0.99030646 which is pretty close and indeed a satisfactory output.